### ISSUE TITLE
omfwd: udp sender does not handle  EAGAIN

### ISSUE THREAD

**Issue by rgerhards** [rsyslog/rsyslog#1665]
On very busy systems, we see "udp send error 11" inside the logs, and the requesting action is being suspended (and later resumed). During the suspension period (in default configuration), messages are lost. Error 11 translates to EAGAIN and the cause of this problem is that the system is running out of UDP buffer space. This can happen on very busy systems (with busy networks).

It is not an error per se. Doing a short wait will resolve the issue. The real root cause of the issue is that omfwd uses a nonblocking socket for sending. If it were blocking, the OS would block until the situation is resolved. The need for a non-blocking sockets is a purely historical one. In the days of single-threaded processing (pre v3), everything needed to be done by multiplexing, and blocking was not permitted. Since then, the engine has dramatically changed. Actions now run on their own thread(s). As such, there is no longer a hard need to use non-blocking i/o for sending data. Many other output plugins also do blocking wait (e.g. omelasticsearch). As such, the real root cause of the trouble is unnecessarily using non-blocking mode, and consequently the right solution is to change that.

Note that using blocking i/o might change some timeing inside rsyslog, especially during shutdown. So theoretical there is regression potential in that area. However, the core is designed to handle that situation (e.g. there is special shutdown code to handle the blocking case), so this does not stand against the "proper" solution.

**Comment by davidelang** [comment]
will blocking IO actually wait for UDP? the OS stack is allowed to drop UDP if 
the network is too busy (per spec)

**Comment by rgerhards** [comment]
Thanks for the comment, @davidelang . I had thought that when EAGAIN is returned in the nonblocking case, a decent wait will occur in the blocking case. But you are right, this might actually be different with UDP. So I investigated.

First of all, we need to be aware that the RFC says UDP can be discarded if there is need to do so. This is why I wondered to get EAGAIN in the first place. Unfortunately, whatever we may find out in the difference between blocking and non-blocking might be highly version dependent, as based on RFC behavior may validly be different.

That said, I first looked at the kernel source tree. sendto() is translated to sendmsg() which later in the process uses a protocol-specific driver which later in the process probably calls some driver function which ultimately defines the behavior. I did review code for roughly half an hour without coming to a final conclusion.

Then I started by googling a bit, hoping to get a faster response. I found two very interesting post, one of which I would like to partly reproduce here:

```With "some" (sorry) Linux versions, you must have SO_SNDBUF lower than
txqueuelen if you want to avoid dropping UDP packets. Yes, I wrote
LOWER, not bigger. That is because the txqueue silently drops packets
when it overflows (this is not a bug: IP is NOT reliable). On the
other hand the send() system call blocks the iperf process and does
NOT drop data when the socket buffer is full (probably because send()
does not make any difference between UDP, TCP and others).
```
This is reproduced from http://www.mail-archive.com/iperf-users@lists.sourceforge.net/msg00113.html and referred to by https://stackoverflow.com/questions/4165174/when-does-a-udp-sendto-block. The latter also has information that matches what I initially wrote:

```This can happen if you filled up your socket buffer, but it is highly operating 
system dependent. Since UDP does not provide any guarantee your operating
system can decide to do whatever it wants when your socket buffer is 
full: block or drop. You can try to increase SO_SNDBUF for temporary relief.

This can even depend on the fine tuning of your system, for instance it can also
depend on the size of the TX ring in the driver of your network interface. There are
a few discussions about this in the iperf mailing list, but you really want to discuss
this with the developers of your operating system.
```

After reading this, I went to the kernel source once again. I did not try hard to validate the postings (reasons follow), but I think I can confirm this by looking at function sock_alloc_send_pskb in ./net/core/sock.c (link to current version, might have changed: http://elixir.free-electrons.com/linux/latest/source/net/core/sock.c#L1970): I think this is the function that is used when sendto() tries to put data to the socket buffer. If so, it means a blocking socket will actually wait, probably for the reasons given in the first post.

But even if I picked the wrong function: it looks very evident that the exact behavior cannot be counted on. As such, I assume that that using a blocking socket to avoid the EAGAIN is probably the best solution. The exact semantics depend on OS, it's version and possibly also on other factors. With that level of uncertainty, I think it is not justified to use app-level busy waiting to try to gain additional guarantees. After all, one can argue that EAGAIN needs not to be issued and the message just be discarded by the stack. As it  looks, for current Linux, both approaches also fulfill the same semantics, albeit the blocking case in a more elegant way.

@davidelang does that address your concerns or is there anything else we need to look into?
